File: ecosystem/naalp-agent-go/plan/orchestrator.go
// Copyright (c) 2026 BubbleFish Technologies, Inc. Apache-2.0.

// Package plan is the N-AALP Plan-and-Execute orchestrator for Go agents (ecosystem task
// E1.4, requirement R2.4 -- mirroring the shipped Python ecosystem task E1.2/R2.2).
//
// Given a plan -- a DAG of tasks, each with an id, an Action, and a set of prerequisite
// task ids -- the orchestrator converts each node to a real signed N-AALP request object,
// sets its causal edges (Causes) to the content ids of its prerequisite nodes' request
// objects, executes nodes in a valid topological order, and is fail-closed: a node whose
// prerequisite refused/failed (or whose own request cannot be signed/validated) is NOT
// executed, and a cycle in the plan is rejected with a named error before any signing.
//
//	Plan (Task DAG) -> Orchestrator.Run() -> a Run: per-task NodeResult, in the
//	                                          topological order actually executed.
//
// This package performs NO cryptography, NO CBOR encoding, and NO second content-id
// computation of its own: every signed-object-level operation (canonical encoding,
// content-id binding, deterministic ML-DSA sign/verify, the audience point-of-use gate,
// the causal-linkage check on a response) is the real sibling react.Bridge's own
// ActionToRequest/ResponseToObservation, which in turn is the real Part-1
// impl/go/envelope primitive (design.md "Buy-before-make"; this package is
// DAG-scheduling and fail-closed-propagation glue, not a second copy of the ReAct bridge
// or the envelope/COSE primitive it wraps).
//
// Design choice -- the causal graph here is a PLAN-LEVEL, pre-signing DAG over
// caller-chosen task ids (plain strings), never impl/go/audit.VerifyCausal's SIGNED-object
// causal graph: VerifyCausal reconciles a set of ALREADY-SIGNED, ALREADY-POSITIONED nodes
// (it needs each node's final topological position up front, which does not exist yet for
// a plan that has not executed a single node), so it answers a different question than the
// one this orchestrator asks before any node is signed: "does this caller-declared task-id
// DAG contain a cycle, and in what order can its nodes be executed?" That question is
// answered here by a plain Kahn's-algorithm topological sort (ties broken lexicographically
// by task id, for a deterministic execution order) -- exactly mirroring the sibling
// react package's own documented choice not to reach for a graph-verification primitive
// when the narrower, pre-signing question is a better fit. Once a node IS signed, its
// causal edges are real N-AALP Causes entries -- the content ids of its prerequisites'
// real signed request objects -- so the resulting signed graph is exactly what
// audit.VerifyCausal would itself accept were it later run over the finished, positioned
// output (§8.2-§8.3): a plan-level topological sort here does not compete with that
// grading; it produces input that graph checks are already known to accept.
//
// Design choice -- an injected, interface-typed Bridge (never a hard construction of
// one): Bridge is any type exposing ActionToRequest(Action) (Request, error) and
// ResponseToObservation([]byte, Request) Observation -- exactly the sibling
// react.Bridge's own two public methods, so a real *react.Bridge (with its own
// clock/signing-key/audience/transport already set on IT) plugs in with zero adaptation,
// and a test can substitute a thin recording wrapper around a real bridge instead.
// Executor is any func(Task, Request) ([]byte, error): the synchronous "send this signed
// request, get the raw response bytes back" step -- the orchestrator's own analogue of a
// transport, injected exactly like react.Bridge's own Transport field, so no node's
// request/response cycle is ever driven by a hidden network call or a global. This keeps
// the whole orchestrator testable in isolation (this task's isolation-demo requirement)
// while still composing with a real bridge and a real transport exactly as the design
// intends.
//
// Design choice -- fail-closed by CATCHING, not by requiring callers to pre-filter: Run
// never lets a per-node signing/validation/execution/verification failure escape as a
// panic that aborts the whole plan (Go's analogue of Python's per-call try/except). Each
// of those three failure points is checked at its own call site and turned into a named
// NodeResult{Status: "refused", ...} for exactly that node, so one bad node degrades
// gracefully into "this node and everything that depends on it did not run" rather than
// crashing every other, unrelated branch of the plan. A cycle is the one failure that IS
// returned as an error from Run itself (*Error{"CyclicPlan", ...}) rather than recorded
// per-node, because a cyclic plan has no valid execution order at all -- there is no
// partial result to report, and it must be rejected before any node's request is ever
// signed.
package plan

import (
	"fmt"
	"sort"
	"strings"

	"github.com/bubblefish-tech/naalp_protocol/ecosystem/naalp-agent-go/react"
)

// Error is a named, fail-closed plan-structure error (E1.4/R2.4). Kind is one of:
// "CyclicPlan" (the task DAG contains a cycle -- returned before any node is signed),
// "DuplicateTaskId" (two tasks in the same plan share an id), or "UnknownPrerequisite"
// (a task names a prerequisite id that is not itself a task in the plan).
type Error struct {
	Kind string
	Msg  string
}

func (e *Error) Error() string { return e.Kind + ": " + e.Msg }

// Task is one node in a plan (R2.4's "a DAG of tasks, each with a name, an effect class,
// args, and a set of prerequisite task-ids"). ID is the caller-chosen, plan-local
// identifier used only to express the DAG (Prereqs) -- it never reaches the wire. Action
// is the N-AALP react.Action this node will request (name/channel/kind/effect/args); its
// own Causes field is IGNORED and overwritten by the orchestrator with the real content
// ids of Prereqs' signed request objects -- the caller declares causal STRUCTURE via
// Prereqs, never authors Causes bytes directly, because those bytes do not exist until a
// prerequisite is actually signed. Prereqs is nil/empty for a first-turn (root) task.
type Task struct {
	ID      string
	Action  react.Action
	Prereqs []string
}

// NodeResult is the outcome of one task after Orchestrator.Run(). Status is exactly one
// of: "executed" (the request was signed, the executor ran, and the response verified
// OK), "refused" (this task's OWN request could not be signed/validated, its executor
// call failed, or its response failed verification), or "blocked" (a prerequisite was
// not "executed", so this task's request was never even built or signed). Request is set
// whenever a signed request was actually produced (including a later-refused response);
// Observation is set only when a response was actually verified (never fabricated on a
// "blocked" or pre-verification "refused" outcome).
type NodeResult struct {
	TaskID      string
	Status      string
	Request     *react.Request
	Observation *react.Observation
	Detail      string
}

// Run is the complete result of executing a plan: the topological order actually used,
// and the per-task NodeResults (one entry per task in the plan, always present regardless
// of outcome).
type Run struct {
	Order      []string
	Results    map[string]NodeResult
	StartedAt  uint64
	FinishedAt uint64
}

// ObservationFor returns the Observation for taskID (nil if none was ever verified).
func (r Run) ObservationFor(taskID string) *react.Observation {
	return r.Results[taskID].Observation
}

// RequestIDFor returns the content id of taskID's signed request (nil if none was ever
// produced).
func (r Run) RequestIDFor(taskID string) []byte {
	if req := r.Results[taskID].Request; req != nil {
		return req.ID
	}
	return nil
}

// Executed returns the task ids that reached Status == "executed", in Run.Order.
func (r Run) Executed() []string {
	out := make([]string, 0, len(r.Order))
	for _, tid := range r.Order {
		if r.Results[tid].Status == "executed" {
			out = append(out, tid)
		}
	}
	return out
}

// Bridge is the interface Orchestrator depends on -- exactly react.Bridge's own two
// public methods, so a real *react.Bridge plugs in with zero adaptation.
type Bridge interface {
	ActionToRequest(action react.Action) (react.Request, error)
	ResponseToObservation(responseBytes []byte, request react.Request) react.Observation
}

// Executor sends task's signed request and returns the raw response bytes.
type Executor func(task Task, request react.Request) ([]byte, error)

// Orchestrator schedules a Task DAG into a valid topological order, converting each task
// into a real signed N-AALP request (via the injected Bridge) whose Causes names exactly
// its prerequisites' signed request content ids, running the injected Executor to get
// each request's response, and verifying that response back into an Observation (via the
// same injected Bridge) before letting any dependent task proceed. The Bridge, the
// Executor, and the Clock are all struct fields the caller sets directly (never a hidden
// global, never a network call this type makes itself, never a bridge this type
// constructs on its own) so the whole orchestrator is testable in isolation.
type Orchestrator struct {
	Bridge   Bridge
	Executor Executor
	// Clock returns the wall clock in epoch milliseconds; react.DefaultClockMS if nil.
	Clock func() uint64
}

// New returns a ready Orchestrator wrapping bridge and executor.
func New(bridge Bridge, executor Executor) *Orchestrator {
	return &Orchestrator{Bridge: bridge, Executor: executor}
}

func (o *Orchestrator) clock() uint64 {
	if o.Clock != nil {
		return o.Clock()
	}
	return react.DefaultClockMS()
}

// Run executes tasks to completion. Returns a non-nil *Error (before any signing) if the
// plan itself is malformed (a cycle, a duplicate id, or an unknown prerequisite);
// otherwise NEVER returns an error -- every per-node failure is recorded as that node's
// own NodeResult{Status: "refused"/"blocked", ...} and execution continues over the rest
// of the plan (package docstring, "fail-closed by catching").
func (o *Orchestrator) Run(tasks []Task) (Run, error) {
	byID, order, err := validateAndOrder(tasks)
	if err != nil {
		return Run{}, err
	}

	startedAt := o.clock()
	results := make(map[string]NodeResult, len(byID))
	requestIDs := make(map[string][]byte, len(byID))

	for _, tid := range order {
		task := byID[tid]

		var unresolved []string
		for _, p := range task.Prereqs {
			if results[p].Status != "executed" {
				unresolved = append(unresolved, p)
			}
		}
		if len(unresolved) > 0 {
			// (task bar c/e): a node whose prerequisite refused, was itself blocked, or
			// never executed is NOT executed -- no request is ever built or signed for
			// it, and no Observation is ever fabricated.
			sort.Strings(unresolved)
			results[tid] = NodeResult{
				TaskID: tid, Status: "blocked",
				Detail: "blocked on unresolved prerequisite(s): " + strings.Join(unresolved, ", "),
			}
			continue
		}

		causes := make([][]byte, 0, len(task.Prereqs))
		for _, p := range task.Prereqs {
			causes = append(causes, requestIDs[p])
		}
		linkedAction := task.Action
		linkedAction.Causes = causes

		request, err := o.Bridge.ActionToRequest(linkedAction)
		if err != nil {
			// fail-closed: ANY sign/validate refusal degrades this node only.
			results[tid] = NodeResult{
				TaskID: tid, Status: "refused",
				Detail: fmt.Sprintf("request signing/validation refused: %v", err),
			}
			continue
		}
		requestIDs[tid] = request.ID

		responseBytes, err := o.Executor(task, request)
		if err != nil {
			// an executor failure refuses this node; it is never an Observation.
			results[tid] = NodeResult{
				TaskID: tid, Status: "refused", Request: &request,
				Detail: fmt.Sprintf("executor failed: %v", err),
			}
			continue
		}

		observation := o.Bridge.ResponseToObservation(responseBytes, request)
		if !observation.OK {
			results[tid] = NodeResult{
				TaskID: tid, Status: "refused", Request: &request, Observation: &observation,
				Detail: "response verification refused: " + observation.Error,
			}
			continue
		}

		results[tid] = NodeResult{TaskID: tid, Status: "executed", Request: &request, Observation: &observation}
	}

	finishedAt := o.clock()
	return Run{Order: order, Results: results, StartedAt: startedAt, FinishedAt: finishedAt}, nil
}

// validateAndOrder builds the id->Task map and computes a deterministic topological
// order, or returns a named *Error -- BEFORE any node is signed. This is the structural
// gate Run depends on: Run calls this before its execution loop ever touches
// Bridge.ActionToRequest, so a cyclic (or otherwise malformed) plan produces zero signed
// bytes.
func validateAndOrder(tasks []Task) (map[string]Task, []string, error) {
	byID := make(map[string]Task, len(tasks))
	for _, t := range tasks {
		if _, exists := byID[t.ID]; exists {
			return nil, nil, &Error{"DuplicateTaskId", fmt.Sprintf("task id %q appears more than once in the plan", t.ID)}
		}
		byID[t.ID] = t
	}

	for _, t := range byID {
		for _, p := range t.Prereqs {
			if _, ok := byID[p]; !ok {
				return nil, nil, &Error{"UnknownPrerequisite", fmt.Sprintf("task %q names unknown prerequisite %q", t.ID, p)}
			}
		}
	}

	order, err := topologicalOrder(byID)
	if err != nil {
		return nil, nil, err
	}
	return byID, order, nil
}

// topologicalOrder runs Kahn's algorithm over the Prereqs DAG, ties broken
// lexicographically by task id for a deterministic execution order. Returns
// *Error{"CyclicPlan", ...} naming every task id that could never be scheduled -- the
// definitive sign of a cycle (a self-loop included: a task naming itself as its own
// prerequisite can never reach indegree 0).
func topologicalOrder(byID map[string]Task) ([]string, error) {
	indegree := make(map[string]int, len(byID))
	dependents := make(map[string][]string, len(byID))
	for tid := range byID {
		indegree[tid] = 0
	}
	for tid, task := range byID {
		for _, p := range task.Prereqs {
			dependents[p] = append(dependents[p], tid)
			indegree[tid]++
		}
	}

	var ready []string
	for tid, deg := range indegree {
		if deg == 0 {
			ready = append(ready, tid)
		}
	}
	sort.Strings(ready)

	order := make([]string, 0, len(byID))
	for len(ready) > 0 {
		sort.Strings(ready)
		current := ready[0]
		ready = ready[1:]
		order = append(order, current)
		for _, dep := range dependents[current] {
			indegree[dep]--
			if indegree[dep] == 0 {
				ready = append(ready, dep)
			}
		}
	}

	if len(order) != len(byID) {
		seen := make(map[string]bool, len(order))
		for _, tid := range order {
			seen[tid] = true
		}
		var unresolved []string
		for tid := range byID {
			if !seen[tid] {
				unresolved = append(unresolved, tid)
			}
		}
		sort.Strings(unresolved)
		return nil, &Error{"CyclicPlan", "the plan's task DAG contains a cycle reachable from: " + strings.Join(unresolved, ", ")}
	}
	return order, nil
}

File: ecosystem/naalp-plan/naalp_plan/orchestrator.py
# Copyright (c) 2026 BubbleFish Technologies, Inc. Apache-2.0.
"""
The N-AALP Plan-and-Execute orchestrator (Part-2 ecosystem task E1.2, requirement R2.2).

requirements.md R2.2: given a plan -- a DAG of tasks, each with a name, an effect class,
args, and a set of prerequisite task ids -- the orchestrator converts each node to a real
signed N-AALP request object, sets its causal edges (`causes[]`) to the content ids of its
prerequisite nodes' request objects, executes nodes in a valid topological order, and is
fail-closed: a node whose prerequisite refused/failed (or whose own request cannot be
signed/validated) is NOT executed, and a cycle in the plan is rejected with a named error
before any signing.

    Plan (Task DAG) -> PlanOrchestrator.run() -> a PlanRun: per-task NodeResult, in the
                                                   topological order actually executed.

This module performs NO cryptography, NO CBOR encoding, and NO second content-id
computation of its own: every signed-object-level operation (canonical encoding,
content-id binding, deterministic ML-DSA sign/verify, the audience point-of-use gate, the
causal-linkage check on a response) is the real sibling `naalp_react.ReActBridge`'s own
`action_to_request` / `response_to_observation`, which in turn is the real Part-1
`naalp.envelope` primitive (design.md "Buy-before-make"; this module is DAG-scheduling and
fail-closed-propagation glue, not a second copy of the ReAct bridge or the envelope/COSE
primitive it wraps).

Design choice -- the causal graph here is a PLAN-LEVEL, pre-signing DAG over caller-chosen
task ids (plain strings), never `naalp.graph.verify_causal`'s SIGNED-object causal graph:
`graph.verify_causal` reconciles a set of ALREADY-SIGNED, ALREADY-POSITIONED nodes (it needs
each node's final topological position up front, which does not exist yet for a plan that
has not executed a single node), so it answers a different question than the one this
orchestrator asks before any node is signed: "does this caller-declared task-id DAG contain
a cycle, and in what order can its nodes be executed?" That question is answered here by a
plain Kahn's-algorithm topological sort (ties broken lexicographically by task id, for a
deterministic execution order), exactly mirroring `naalp_react.bridge`'s own documented
choice not to reach for `graph.verify_causal` when the narrower, pre-signing question is a
better fit (bridge.py's module docstring, "a plain membership check, not
`naalp.graph.verify_causal`"). Once a node IS signed, its causal edges are real N-AALP
`causes[]` entries -- the content ids of its prerequisites' real signed request objects --
so the resulting signed graph is exactly what `naalp.graph.verify_causal` would itself
accept were it later run over the finished, positioned output (§8.2-§8.3): a plan-level
topological sort here does not compete with that grading; it produces input that graph.py's
own no-future-cause + acyclicity checks are already known to accept.

Design choice -- injected, duck-typed `bridge` and `executor` (never a hard construction of
either inside this module): `bridge` is any object exposing
`action_to_request(action) -> Request` and `response_to_observation(bytes, Request) ->
Observation` -- exactly `naalp_react.ReActBridge`'s own two public methods, so a real
`ReActBridge` (with its own clock/signing-key/audience/transport already injected into IT)
plugs in with zero adaptation, and a test can substitute a thin recording wrapper around a
real bridge instead. `executor` is any callable `(Task, Request) -> bytes`: the synchronous
"send this signed request, get the raw response bytes back" step -- the orchestrator's own
analogue of a transport, injected exactly like `ReActBridge`'s own `transport=` constructor
argument, so no node's request/response cycle is ever driven by a hidden network call or a
global. This keeps the whole orchestrator testable in isolation (this task's isolation-demo
requirement) while still composing with a real bridge and a real transport exactly as the
design intends.

Design choice -- fail-closed by CATCHING, not by requiring callers to pre-filter: `run()`
never lets a per-node signing/validation/execution/verification failure escape as an
exception that aborts the whole plan. Each of those four failure points is caught at its
own call site and turned into a named `NodeResult(status="refused", ...)` for exactly that
node, so one bad node degrades gracefully into "this node and everything that depends on it
did not run" rather than crashing every other, unrelated branch of the plan. A cycle is the
one failure that IS raised (as `PlanError("CyclicPlan", ...)`) rather than recorded per-node,
because a cyclic plan has no valid execution order at all -- there is no partial result to
report, and (task bar) it must be rejected before any node's request is ever signed.
"""
from dataclasses import dataclass, field, replace
from typing import Callable, Dict, List, Optional, Sequence, Tuple

from . import _bootstrap  # noqa: F401  (side-effecting import: puts naalp_react + impl/python on sys.path)

from naalp_react import Action, Observation, Request, default_clock_ms  # noqa: F401  (re-exported for callers)


class PlanError(ValueError):
    """A named, fail-closed plan-structure error (E1.2/R2.2). `.kind` is one of:
    "CyclicPlan" (the task DAG contains a cycle -- raised before any node is signed),
    "DuplicateTaskId" (two tasks in the same plan share an id), or "UnknownPrerequisite"
    (a task names a prerequisite id that is not itself a task in the plan)."""

    def __init__(self, kind, msg=""):
        super().__init__("%s: %s" % (kind, msg) if msg else kind)
        self.kind = kind


@dataclass(frozen=True)
class Task:
    """One node in a plan (R2.2's "a DAG of tasks, each with a name, an effect class, args,
    and a set of prerequisite task-ids"). `id` is the caller-chosen, plan-local identifier
    used only to express the DAG (`prereqs`) -- it never reaches the wire. `action` is the
    N-AALP `naalp_react.Action` this node will request (name/channel/kind/effect/args);
    its own `causes` field is IGNORED and overwritten by the orchestrator with the real
    content ids of `prereqs`' signed request objects -- the caller declares causal
    STRUCTURE via `prereqs`, never authors `causes[]` bytes directly, because those bytes
    do not exist until a prerequisite is actually signed. `prereqs` is empty for a
    first-turn (root) task."""

    id: str
    action: Action
    prereqs: Tuple[str, ...] = field(default_factory=tuple)


@dataclass(frozen=True)
class NodeResult:
    """The outcome of one task after `PlanOrchestrator.run()`. `status` is exactly one of:
    "executed" (the request was signed, the executor ran, and the response verified `ok`),
    "refused" (this task's OWN request could not be signed/validated, its executor call
    failed, or its response failed verification), or "blocked" (a prerequisite was not
    "executed", so this task's request was never even built or signed). `request` is set
    whenever a signed request was actually produced (including a later-refused response);
    `observation` is set only when a response was actually verified (never fabricated on a
    "blocked" or pre-verification "refused" outcome)."""

    task_id: str
    status: str
    request: Optional[Request] = None
    observation: Optional[Observation] = None
    detail: str = ""


@dataclass(frozen=True)
class PlanRun:
    """The complete result of executing a plan: the topological order actually used, and
    the per-task `NodeResult`s (task_id -> NodeResult, one entry per task in the plan,
    always present regardless of outcome)."""

    order: Tuple[str, ...]
    results: Dict[str, NodeResult]
    started_at: int
    finished_at: int

    def observation_for(self, task_id: str) -> Optional[Observation]:
        return self.results[task_id].observation

    def request_id_for(self, task_id: str) -> Optional[bytes]:
        r = self.results[task_id].request
        return r.id if r is not None else None

    def executed(self) -> Tuple[str, ...]:
        return tuple(tid for tid in self.order if self.results[tid].status == "executed")


Executor = Callable[[Task, Request], bytes]


def _validate_and_order(tasks: Sequence[Task]) -> Tuple[Dict[str, Task], Tuple[str, ...]]:
    """Build the id->Task map and compute a deterministic topological order, or raise a
    named `PlanError` -- BEFORE any node is signed. This is the structural gate task bar
    (d) depends on: `PlanOrchestrator.run()` calls this before its execution loop ever
    touches `bridge.action_to_request`, so a cyclic (or otherwise malformed) plan produces
    zero signed bytes."""
    tasks_by_id: Dict[str, Task] = {}
    for t in tasks:
        if t.id in tasks_by_id:
            raise PlanError("DuplicateTaskId", "task id %r appears more than once in the plan" % (t.id,))
        tasks_by_id[t.id] = t

    for t in tasks_by_id.values():
        for p in t.prereqs:
            if p not in tasks_by_id:
                raise PlanError(
                    "UnknownPrerequisite", "task %r names unknown prerequisite %r" % (t.id, p)
                )

    order = _topological_order(tasks_by_id)
    return tasks_by_id, order


def _topological_order(tasks_by_id: Dict[str, Task]) -> Tuple[str, ...]:
    """Kahn's algorithm over the prereqs DAG, ties broken lexicographically by task id for
    a deterministic execution order. Raises `PlanError("CyclicPlan", ...)` naming every
    task id that could never be scheduled -- the definitive sign of a cycle (a self-loop
    included: a task naming itself as its own prerequisite can never reach indegree 0)."""
    indegree: Dict[str, int] = {tid: 0 for tid in tasks_by_id}
    dependents: Dict[str, List[str]] = {tid: [] for tid in tasks_by_id}
    for tid, task in tasks_by_id.items():
        for p in task.prereqs:
            dependents[p].append(tid)
            indegree[tid] += 1

    ready = sorted(tid for tid, deg in indegree.items() if deg == 0)
    order: List[str] = []
    while ready:
        ready.sort()
        current = ready.pop(0)
        order.append(current)
        for dep in dependents[current]:
            indegree[dep] -= 1
            if indegree[dep] == 0:
                ready.append(dep)

    if len(order) != len(tasks_by_id):
        unresolved = sorted(set(tasks_by_id) - set(order))
        raise PlanError(
            "CyclicPlan", "the plan's task DAG contains a cycle reachable from: %s" % ", ".join(unresolved)
        )
    return tuple(order)


class PlanOrchestrator:
    """R2.2: schedule a Task DAG into a valid topological order, converting each task into
    a real signed N-AALP request (via the injected `bridge`) whose `causes[]` names exactly
    its prerequisites' signed request content ids, running the injected `executor` to get
    each request's response, and verifying that response back into an `Observation` (via
    the same injected `bridge`) before letting any dependent task proceed. The bridge, the
    executor, and the clock are all constructor-injected (never a hidden global, never a
    network call this class makes itself, never a bridge this class constructs on its own)
    so the whole orchestrator is testable in isolation."""

    def __init__(self, *, bridge, executor: Executor, clock: Callable[[], int] = default_clock_ms):
        self._bridge = bridge
        self._executor = executor
        self._clock = clock

    def run(self, tasks: Sequence[Task]) -> PlanRun:
        """Execute `tasks` to completion. Raises `PlanError` (before any signing) if the
        plan itself is malformed (a cycle, a duplicate id, or an unknown prerequisite);
        otherwise NEVER raises -- every per-node failure is recorded as that node's own
        `NodeResult(status="refused"/"blocked", ...)` and execution continues over the
        rest of the plan (module docstring, "fail-closed by catching")."""
        tasks_by_id, order = _validate_and_order(tasks)

        started_at = self._clock()
        results: Dict[str, NodeResult] = {}
        request_ids: Dict[str, bytes] = {}

        for tid in order:
            task = tasks_by_id[tid]

            unresolved = [p for p in task.prereqs if results[p].status != "executed"]
            if unresolved:
                # (task bar c/e): a node whose prerequisite refused, was itself blocked, or
                # never executed is NOT executed -- no request is ever built or signed for
                # it, and no Observation is ever fabricated.
                results[tid] = NodeResult(
                    task_id=tid,
                    status="blocked",
                    detail="blocked on unresolved prerequisite(s): %s" % ", ".join(sorted(unresolved)),
                )
                continue

            linked_action = replace(task.action, causes=tuple(request_ids[p] for p in task.prereqs))

            try:
                request = self._bridge.action_to_request(linked_action)
            except Exception as e:  # noqa: BLE001 -- fail-closed: ANY sign/validate refusal degrades this node only
                results[tid] = NodeResult(
                    task_id=tid,
                    status="refused",
                    detail="request signing/validation refused: %s: %s" % (type(e).__name__, e),
                )
                continue

            request_ids[tid] = request.id

            try:
                response_bytes = self._executor(task, request)
            except Exception as e:  # noqa: BLE001 -- an executor failure refuses this node; it is never an Observation
                results[tid] = NodeResult(
                    task_id=tid, status="refused", request=request,
                    detail="executor failed: %s: %s" % (type(e).__name__, e),
                )
                continue

            observation = self._bridge.response_to_observation(response_bytes, request)
            if not observation.ok:
                results[tid] = NodeResult(
                    task_id=tid, status="refused", request=request, observation=observation,
                    detail="response verification refused: %s" % observation.error,
                )
                continue

            results[tid] = NodeResult(task_id=tid, status="executed", request=request, observation=observation)

        finished_at = self._clock()
        return PlanRun(order=order, results=results, started_at=started_at, finished_at=finished_at)
